### PR TITLE
Python QPS Worker/initial scenarios

### PR DESCRIPTION
**Pull request by kpayson64**
Created python QPS worker and Jenkins integration #4178

### CODE REVIEW

File: tools/run_tests/performance/scenario_config.py [hunk 1]
```diff
@@ -428,6 +428,83 @@ def scenarios(self):
...
+class PythonLanguage:
+
+  def __init__(self):
+    self.safename = 'python'
+
+  def worker_cmdline(self):
+    return ['tools/run_tests/performance/run_worker_python.sh']
+
+  def worker_port_offset(self):
+    return 300
```
Comment (jtattermusch) [inline tools/run_tests/performance/scenario_config.py]: pls use 500 (300 is used by ruby and 400 for java is coming).

File: src/python/grpcio/tests/qps/control.proto [hunk 2]
```diff
@@ -0,0 +1,196 @@
+// Copyright 2015, Google Inc.
```
Comment (jtattermusch) [inline src/python/grpcio/tests/qps/control.proto]: we are trying very hard not to copy the .proto files around (we've had enough problems with having multiple copies of the same .proto files), so the protos in src/proto should be used.

Comment (jtattermusch) [inline src/python/grpcio/tests/qps/control.proto]: same for all other .proto files.

File: tools/run_tests/performance/scenario_config.py [hunk 3]
```diff
@@ -476,6 +553,7 @@ def __str__(self):
 
 
 LANGUAGES = {
+    'python' : PythonLanguage(),
```
Comment (jtattermusch) [inline tools/run_tests/performance/scenario_config.py]: nit: let's use alphabetical order or add python at the end.

File: src/python/grpcio/tests/qps/qps_worker.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+./qps_worker.py --driver_port <port>
+"""
+
+def run_worker_server(port):
+  servicer = WorkerServer()
+  server = services_pb2.beta_create_WorkerService_server(servicer)
+  server.add_insecure_port('[::]:{}'.format(port))
+  server.start()
+  servicer.wait_for_quit()
+  time.sleep(2)
```
Comment (jtattermusch) [inline src/python/grpcio/tests/qps/qps_worker.py]: in the previous PR you've mentioned you're using this as a workaround. If so, please add and explanatory comment (or TODO).

File: src/python/grpcio/tests/qps/histogram.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import math
+import threading
+
+from tests.qps.stats_pb2 import HistogramData
+ 
+"""Histogram class used for recording performance testing data.
```
Comment (jtattermusch) [inline src/python/grpcio/tests/qps/histogram.py]: The comment goes after the class declaration.
We are trying to follow the go/python-style python style where possible.

File: src/python/grpcio/tests/qps/benchmark_client.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+"""Abstract benchmark client interface that exposes a non-blocking 
+send_request() method used by the client runners
+"""
+class BenchmarkClient:
+  __metaclass__ = abc.ABCMeta
+
+
+  def __init__(self, server, payload_config, use_ssl, hist):    
+    #Create the stub
```
Comment (jtattermusch) [inline src/python/grpcio/tests/qps/benchmark_client.py]: space after #

File: src/python/grpcio/tests/qps/benchmark_client.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+from tests.unit import resources
+from tests.unit.beta import test_utilities
+
+from tests.interop.messages_pb2 import Payload, SimpleRequest 
+from tests.qps import services_pb2
+
+_TIMEOUT = 60 * 60 * 24
+_SERVER_HOST_OVERRIDE = 'foo.test.google.fr'
+
+"""Abstract benchmark client interface that exposes a non-blocking
```
Comment (jtattermusch) [inline src/python/grpcio/tests/qps/benchmark_client.py]: comment goes after class declaration

**Comment by jtattermusch** [comment]
Please let me know once this is ready for review.

**Comment by kpayson64** [comment]
Because this might be pending until the proto build changes go through, here are the initial performance numbers running against the c++ server: 

Streaming client:
```
python_protobuf_async_streaming_ping_pong_secure
I0422 14:34:23.518312206   39676 report.cc:77]               QPS: 287.4
I0422 14:34:23.520832590   39676 report.cc:81]               QPS: 287.4 (23.9/server core)
I0422 14:34:23.520851201   39676 report.cc:86]               Latencies (50/90/95/99/99.9%-ile): 3403.8/3701.8/3813.2/4087.1/5533.4 us
```

Sync Unary Client
```
python_protobuf_sync_unary_ping_pong_secure
 QPS: 222.4
I0422 14:34:58.584801337   39716 report.cc:81]               QPS: 222.4 (18.5/server core)
I0422 14:34:58.584832061   39716 report.cc:86]               Latencies (50/90/95/99/99.9%-ile): 4169.5/4829.8/5041.6/5580.3/6549.9 us
```

Async Unary client
```
python_protobuf_async_unary_ping_pong_secure
QPS: 286.8
I0422 14:35:33.652040385   47548 report.cc:81]               QPS: 286.8 (23.9/server core)
I0422 14:35:33.652070038   47548 report.cc:86]               Latencies (50/90/95/99/99.9%-ile): 3445.8/4020.2/4190.8/4550.9/5265.7 us
```

**Comment by kpayson64** [comment]
This latest push has a dependency on  #6229 for proto generation.

**Comment by kpayson64** [comment]
@ctiller 
Throughput test result:
```
I0422 17:59:07.686535969   55344 report.cc:77]               QPS: 64.7
I0422 17:59:07.688609458   55344 report.cc:81]               QPS: 64.7 (16.2/server core)
I0422 17:59:07.688640256   55344 report.cc:86]               Latencies (50/90/95/99/99.9%-ile): 15491.3/15793.6/15971.6/16976.8/24052.6 us
```

**Comment by vjpai** [comment]
Those numbers look like the unconstrained test is thrashing. We probably
need a better selection for the values of wide, deep, and big.

**Comment by kpayson64** [comment]
Moved to #6298

**Pull request closed without merging**